table/tables/cache.go

```diff
@@ -30,74 +34,148 @@ var _ table.CachedTable = &cachedTable{}
 
 type cachedTable struct {
 	TableCommon
-	kv.MemBuffer
-	mu sync.RWMutex
+	cacheData atomic.Value
+	handle    StateRemote
+}
+
+func leaseFromTS(ts uint64) uint64 {
+	// TODO make this configurable in the following PRs
+	const defaultLeaseDuration time.Duration = 3 * time.Second
+	physicalTime := oracle.GetTimeFromTS(ts)
+	lease := oracle.GoTimeToTS(physicalTime.Add(defaultLeaseDuration))
+	return lease
 }
 
 func (c *cachedTable) TryGetMemcache(ts uint64) (kv.MemBuffer, bool) {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-	if c.isReadFromCache(ts) {
-		return c.MemBuffer, true
+	tmp := c.cacheData.Load()
+	if tmp == nil {
+		return nil, false
+	}
+	data := tmp.(*table.CacheData)
+	if data.Lease > ts {
+		return data.MemBuffer, true
 	}
 	return nil, false
 }
-func (c *cachedTable) isReadFromCache(ts uint64) bool {
-	// If first read cache table. directly return false, the backend goroutine will help us update the lock information
-	// and read the data from the original table at the same time
-	// TODO : Use lease and ts judge whether it is readable.
-	// TODO : If the cache is not readable. MemBuffer become invalid.
-	return c.MemBuffer != nil
-}
+
+var mockStateRemote = struct {
+	Ch   chan remoteTask
+	Data *mockStateRemoteData
+}{}
 
 // NewCachedTable creates a new CachedTable Instance
 func NewCachedTable(tbl *TableCommon) (table.Table, error) {
-	return &cachedTable{
+	if mockStateRemote.Data == nil {
+		mockStateRemote.Data = newMockStateRemoteData()
+		mockStateRemote.Ch = make(chan remoteTask, 100)
+		go mockRemoteService(mockStateRemote.Data, mockStateRemote.Ch)
+	}
+	ret := &cachedTable{
 		TableCommon: *tbl,
-	}, nil
+		handle:      &mockStateRemoteHandle{mockStateRemote.Ch},
+	}
+
+	return ret, nil
 }
 
-func (c *cachedTable) loadDataFromOriginalTable(ctx sessionctx.Context) error {
+func (c *cachedTable) loadDataFromOriginalTable(ctx sessionctx.Context, lease uint64) (kv.MemBuffer, error) {
 	prefix := tablecodec.GenTablePrefix(c.tableID)
 	txn, err := ctx.Txn(true)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	if txn.StartTS() >= lease {
```

Comment:
You can start a new txn with the same startTs with the old one to avoid this error

Reply:
I agree to implement it in next pr. But I think `txn.RunInNewTxn` is not allowed to specify a custom ts.

Reply:
Use `txn.RunInNewTxn` and compare its `ts` with `lease`,  if `ts < lease`, caching is save. Otherwise we can return a error to the caller